### PR TITLE
Async Halo2 verifier service

### LINKED ISSUE #2104: Implement a `Halo2Verifier` async service
Similar to our other async `tower`-based verifier services, such as the [`Groth16Verifier`](https://github.com/ZcashFoundation/zebra/blob/main/zebra-consensus/src/primitives/groth16.rs).  We should depend on the same version of the `halo2` crate as we do in `zebra-chain`. 

While we don't have batch math available yet for Halo2, the async service will decouple the verification requests from the fullfillment, allowing us to swap out the original single-verification implementation here with a more efficient batch verification implementation that is oblivious at the API level.

**Testing**:

Generate test vectors for halo2 proofs from zcash-hackworks; if they give trouble in our full validation pipeline, punt this full  integration until we get the first NU5 block with an Action in it for testnet.

**Additional context**
#1951

### PR DESCRIPTION
**Pull request by dconnolly**
## Motivation

To match our verification patterns, we need a Halo2 proof async verifier service.

## Solution

Implements a Halo2 verifier service that reproduces the batch api, to maintain the same API as out other verifier services and to be future-proof for batch optimizations for Halo2.

We don't have V5 transactions with verifiable Action proofs yet to use as test vectors, when we can mint those via zcashd they should be used to test this.

In the meantime, we produce test-specific Halo2 proofs. We should augment with test vectors as soon as they are available. 

Resolves https://github.com/ZcashFoundation/zebra/issues/2104

## Review

Anyone but especially @jvff @conradoplg @upbqdn @teor2345 

### Reviewer Checklist

  - [x] Code implements Specs and Designs
  - [x] Tests for Expected Behaviour
  - [x] Tests for Errors

## Follow Up Work

- [ ] When finalized, encode the hash of the `VerifyingKey` and assert that it matches the `::build()`'d one

### CODE REVIEW

**Comment by dconnolly** [comment]
Now unblocked as the orchard crate/circuit has been released

**Review by teor2345 — COMMENTED** [review]
I've skimmed this PR and I can't see anything particularly risky or concerning here.

I'd like to leave the detailed review to other people.

File: zebra-consensus/src/primitives/halo2.rs [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+            .map(|action| {
+                circuit::Instance::from_parts(
+                    anchor,
+                    value::ValueCommitment::from_bytes(&action.cv.into()).unwrap(),
+                    note::Nullifier::from_bytes(&action.nullifier.into()).unwrap(),
+                    redpallas::VerificationKey::<redpallas::SpendAuth>::try_from(<[u8; 32]>::from(
+                        action.rk,
+                    ))
+                    .unwrap(),
+                    note::ExtractedNoteCommitment::from_bytes(&action.cm_x.into()).unwrap(),
```
Comment (conradoplg) [inline zebra-consensus/src/primitives/halo2.rs]: I suggest using `expect` instead of `unwrap` with the reason it's OK to ignore errors (because we already validated them when parsing the tx, right?)

File: zebra-consensus/src/primitives/halo2.rs [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+
+                            result
+                        }
+                        Err(RecvError::Lagged(_)) => {
+                            tracing::error!(
+                                "missed channel updates, BROADCAST_BUFFER_SIZE is too low!!"
+                            );
+                            // This is the enum variant that
+                            // orchard::circuit::Proof.verify() returns on
+                            // evaulation failure.
```
Comment (conradoplg) [inline zebra-consensus/src/primitives/halo2.rs]:
```suggestion
                            // evaluation failure.
```

File: zebra-consensus/src/primitives/halo2/tests.rs [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+        let rsp = verifier.ready().await?.call(Item::from(sd));
+
+        async_checks.push(rsp);
+    }
+
+    while let Some(result) = async_checks.next().await {
+        tracing::trace!(?result);
+        result?;
+    }
```
Comment (conradoplg) [inline zebra-consensus/src/primitives/halo2/tests.rs]: Suggestion: test if verification fails with a corrupted input or proof, to make sure we're not accidentally always returning success.

**Review by conradoplg — CHANGES_REQUESTED** [review]
Looks good! I just left some suggestions

File: zebra-consensus/src/primitives/halo2.rs [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+            .actions()
+            .map(|action| {
+                circuit::Instance::from_parts(
+                    anchor,
+                    value::ValueCommitment::from_bytes(&action.cv.into()).unwrap(),
+                    note::Nullifier::from_bytes(&action.nullifier.into()).unwrap(),
+                    redpallas::VerificationKey::<redpallas::SpendAuth>::try_from(<[u8; 32]>::from(
+                        action.rk,
+                    ))
+                    .unwrap(),
```
Comment (jvff) [inline zebra-consensus/src/primitives/halo2.rs]: Out of scope for this PR, but it would be cool to have `impl<T> TryFrom<redpallas::VerificationKeyBytes<T>> for redpallas::VerificationKey<T>` or something, just to optimize the conversions here :laughing:

File: zebra-consensus/src/primitives/halo2.rs (hunk 1 above)
Comment (jvff) [inline zebra-consensus/src/primitives/halo2.rs:102]: Optional: should these two `unwrap`s be `expect`?

File: zebra-consensus/src/primitives/halo2.rs [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+    /// The constraint system is not satisfied.
+    // #[default]
+    ConstraintSystemFailure,
+    /// Catchall for now until https://github.com/zcash/halo2/pull/394 is merged
+    Other,
+}
+
+impl fmt::Display for Halo2Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        write!(f, "{}", self)
```
Comment (jvff) [inline zebra-consensus/src/primitives/halo2.rs]: Isn't this an infinite loop?

**Review by jvff — COMMENTED** [review]
Haven't fully finished reviewing, but I wanted to post a question so I don't block you :sweat_smile:

File: zebra-consensus/src/primitives/halo2/tests.rs [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+async fn verify_orchard_halo2_proofs<V>(
+    verifier: &mut V,
+    shielded_data: Vec<ShieldedData>,
+) -> Result<(), V::Error>
+where
+    V: tower::Service<Item, Response = ()>,
+    <V as tower::Service<Item>>::Error: std::convert::From<
+        std::boxed::Box<dyn std::error::Error + std::marker::Send + std::marker::Sync>,
+    >,
```
Comment (jvff) [inline zebra-consensus/src/primitives/halo2/tests.rs]: Optional suggestion: I think this could be shortened by leveraging Rust's prelude:

```rust
    <V as tower::Service<Item>>::Error: From<Box<dyn std::error::Error + Send + Sync>>,
```
It can be shortened further using `BoxError` (either from `crate` or `tower`):

```suggestion
    <V as tower::Service<Item>>::Error: From<tower::BoxError>,
```

File: zebra-consensus/src/primitives/halo2/tests.rs [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+        tower::service_fn(
+            (|item: Item| ready(item.verify_single(&VERIFYING_KEY).map_err(Halo2Error::from)))
+                as fn(_) -> _,
+        ),
+    );
+
+    // This should fail if any of the proofs fail to validate.
+    verify_orchard_halo2_proofs(&mut verifier, shielded_data)
+        .await
+        .unwrap()
```
Comment (jvff) [inline zebra-consensus/src/primitives/halo2/tests.rs]: Suggestion: I usually use at least one assertion to mark what the test is focused on checking.

```suggestion
    assert!(verify_orchard_halo2_proofs(&mut verifier, shielded_data)
        .await
        .is_ok());
```

**Review by jvff — CHANGES_REQUESTED** [review]
Looks great! I added some small suggestions, but they're optional. The only thing I'd like to be sure before approving is if the `Display` implementation is correct.

File: zebra-consensus/src/primitives/halo2.rs (hunk 4 above)
Comment (dconnolly) [reply zebra-consensus/src/primitives/halo2.rs]: It would yeah 😻

File: zebra-consensus/src/primitives/halo2.rs (hunk 1 above)
Comment (dconnolly) [reply zebra-consensus/src/primitives/halo2.rs:102]: They probably should yeah

File: zebra-consensus/src/primitives/halo2/tests.rs (hunk 6 above)
Comment (dconnolly) [reply zebra-consensus/src/primitives/halo2/tests.rs]: Thanks! Done in 524b964f

File: zebra-consensus/src/primitives/halo2.rs (hunk 1 above)
Comment (dconnolly) [reply zebra-consensus/src/primitives/halo2.rs:102]: Unfortunately the `orchard::note::ExtractNoteCommitment::from_bytes()` returns a [`subtle::CtOption`](https://docs.rs/subtle/2.1.1/subtle/struct.CtOption.html) which doesn't have `expect()`, so a variant of `unwrap` will happen somewhere

Comment (dconnolly) [reply zebra-consensus/src/primitives/halo2.rs:102]: Some changes in 031836ed

File: zebra-consensus/src/primitives/halo2.rs (hunk 4 above)
Comment (dconnolly) [reply zebra-consensus/src/primitives/halo2.rs]: Well, that exists, the problem is these are two different implementations of `redpallas`: the one in `zebra_chain` and the fork of `redjubjub` that  is re-exported by `orchard`, and looks almost the exact same 😅

File: zebra-consensus/src/primitives/halo2.rs (hunk 1 above)
Comment (dconnolly) [reply zebra-consensus/src/primitives/halo2.rs]: Updated some here: 031836edf96d6967234dcb4086474ff3cca32a01

File: zebra-consensus/src/primitives/halo2.rs (hunk 5 above)
Comment (dconnolly) [reply zebra-consensus/src/primitives/halo2.rs]: You're right!

image

image

Comment (dconnolly) [reply zebra-consensus/src/primitives/halo2.rs]: Fixed in 16b2c2b2

File: Cargo.toml [hunk 8]
```diff
@@ -25,6 +25,7 @@ panic = "abort"
 # TODO: remove these after a new librustzcash release.
 # These are librustzcash requirements specified in its workspace Cargo.toml that we must replicate here
 incrementalmerkletree = { git = "https://github.com/zcash/incrementalmerkletree", rev = "b7bd6246122a6e9ace8edb51553fbf5228906cbb" }
-orchard = { git = "https://github.com/zcash/orchard.git", rev = "2c8241f25b943aa05203eacf9905db117c69bd29" }
+# TODO: replace with upstream orchard when these changes are merged
+orchard = { git = "https://github.com/dconnolly/orchard.git", rev = "568e24cd5f129158375d7ac7d98c89ebff4f982f" }
```
Comment (teor2345) [inline Cargo.toml]: Do we want to merge this PR without the upstream orchard changes getting merged?

If we do, let's:
* push the commit we're depending on to a `ZcashFoundation/orchard` fork
* update the dependencies
* remove `dconnolly/orchard` from Zebra's allowed sources
* open a ticket to remove the forked `orchard` in the next sprint (because that's when the orchard release should be cut)

File: zebra-consensus/Cargo.toml [hunk 9]
```diff
@@ -15,9 +15,11 @@ bellman = "0.10.0"
 bls12_381 = "0.5.0"
 chrono = "0.4.19"
 displaydoc = "0.2.2"
+halo2 = "=0.1.0-beta.1"
 jubjub = "0.7.0"
 lazy_static = "1.4.0"
 once_cell = "1.8"
+orchard = "0.0"
```
Comment (teor2345) [inline zebra-consensus/Cargo.toml]: Nit: depend on patch releases.

```suggestion
orchard = "0.0.0"
```

File: zebra-consensus/src/primitives/halo2.rs [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+}
+
+impl fmt::Display for Halo2Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        match self {
+            Halo2Error::ConstraintSystemFailure => write!(f, "Constraint system failure"),
+            _ => write!(f, "Unknown Halo2 error"),
+        }
+    }
+}
```
Comment (jvff) [inline zebra-consensus/src/primitives/halo2.rs]: Optional idea: sorry to bring this up again, but I realized we could have the `Display` implementation generated by `thiserror`.

```suggestion
pub enum Halo2Error {
    /// The constraint system is not satisfied.
    #[error("Constraint system failure")]
    ConstraintSystemFailure,
    
    /// Catchall for now until https://github.com/zcash/halo2/pull/394 is merged
    #[error("Unknown Halo2 error")]
    Other,
}
```

**Review by jvff — DISMISSED** [review]

File: zebra-consensus/src/primitives/halo2.rs (hunk 10 above)
Comment (dconnolly) [reply zebra-consensus/src/primitives/halo2.rs]: My suggestions finally got merged in https://github.com/zcash/halo2/pull/338/files#diff-ce29ac47e4ad4e869b74fd280b52b536e4967606470ae4f64261c50c32e4404f, so if all the other dependencies work out, I might be able to use `halo2::plonk::Error` after all - if not, I'll use your suggestions

File: zebra-consensus/src/primitives/halo2/tests.rs (hunk 3 above)
Comment (dconnolly) [reply zebra-consensus/src/primitives/halo2/tests.rs]: Done in  ea16b947d69c485e9aa1889fe205281ee4f2e044

File: zebra-consensus/src/primitives/halo2.rs (hunk 10 above)
Comment (dconnolly) [reply zebra-consensus/src/primitives/halo2.rs]: Done in 25082b7cdd08ac5489225ee1f742f01bb549d9f7 for now

File: Cargo.toml (hunk 8 above)
Comment (dconnolly) [reply Cargo.toml]: > * push the commit we're depending on to a `ZcashFoundation/orchard` fork

Done!  https://github.com/ZcashFoundation/orchard/tree/patch-1

> * update the dependencies

Done! f9d4b78b

> * remove `dconnolly/orchard` from Zebra's allowed sources

Done! f9d4b78b

> * open a ticket to remove the forked `orchard` in the next sprint (because that's when the orchard release should be cut)

Done! https://github.com/ZcashFoundation/zebra/issues/3056

Comment (dconnolly) [reply Cargo.toml]: And a link in the top level Cargo.toml

f7bf21a2

File: deny.toml [hunk 11]
```diff
@@ -98,8 +98,7 @@ allow-git = [
     "https://github.com/str4d/redjubjub",
 
     # ticket #2523: replace unmaintained multiset
-    "https://github.com/jmitchell/multiset",
-]
+    "https://github.com/jmitchell/multiset",]
```
Comment (jvff) [inline deny.toml]: Nit: formatting

```suggestion
    "https://github.com/jmitchell/multiset",   
]
```

File: zebra-consensus/src/primitives/halo2/tests.rs [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+            crate::primitives::MAX_BATCH_LATENCY,
+        ),
+        tower::service_fn(
+            (|item: Item| ready(item.verify_single(&VERIFYING_KEY).map_err(Halo2Error::from)))
+                as fn(_) -> _,
+        ),
+    );
+
+    println!("Display: {}", Halo2Error::Other);
+
```
Comment (jvff) [inline zebra-consensus/src/primitives/halo2/tests.rs]: Minor: Leftover debug code?

```suggestion
```

File: zebra-consensus/src/primitives/halo2/tests.rs [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+}
+
+async fn verify_orchard_halo2_proofs<V>(
+    verifier: &mut V,
+    shielded_data: Vec<ShieldedData>,
+) -> Result<(), V::Error>
+where
+    V: tower::Service<Item, Response = ()>,
+    <V as tower::Service<Item>>::Error: From<tower::BoxError>,
+    <V as tower::Service<Item>>::Error: std::fmt::Debug,
```
Comment (jvff) [inline zebra-consensus/src/primitives/halo2/tests.rs]: Minor: I thought it might make sense to make it explicit that these are constraints on the same type, to avoid risk of confusion:

```suggestion
    <V as tower::Service<Item>>::Error: From<tower::BoxError> + std::fmt::Debug,
```

File: zebra-consensus/src/primitives/halo2/tests.rs [hunk 14]
```diff
@@ -0,0 +1,240 @@
...
+}
+
+async fn verify_invalid_orchard_halo2_proofs<V>(
+    verifier: &mut V,
+    shielded_data: Vec<ShieldedData>,
+) -> Result<(), V::Error>
+where
+    V: tower::Service<Item, Response = ()>,
+    <V as tower::Service<Item>>::Error: From<tower::BoxError>,
+    <V as tower::Service<Item>>::Error: std::fmt::Debug,
```
Comment (jvff) [inline zebra-consensus/src/primitives/halo2/tests.rs]: Minor: same as above:

```suggestion
    <V as tower::Service<Item>>::Error: From<tower::BoxError> + std::fmt::Debug,
```

File: deny.toml (hunk 11 above)
Comment (teor2345) [inline deny.toml]: Nit: unnecessary whitespace change

```suggestion
    "https://github.com/jmitchell/multiset",
]
```

File: zebra-chain/Cargo.toml [hunk 15]
```diff
@@ -33,7 +33,7 @@ hex = "0.4"
 incrementalmerkletree = "0.1.0"
 jubjub = "0.7.0"
 lazy_static = "1.4.0"
-orchard = { git = "https://github.com/zcash/orchard.git", rev = "2c8241f25b943aa05203eacf9905db117c69bd29" }
+orchard = "0.0"
```
Comment (teor2345) [inline zebra-chain/Cargo.toml:36]: Nit: Let's try to use full patch versions for all our dependencies, to avoid accidental duplicates.

```suggestion
orchard = "0.0.0"
```

File: zebra-consensus/src/primitives/halo2/tests/vectors/orchard_shielded_data.rs [hunk 16]
```diff
@@ -0,0 +1,28 @@
+//! Orchard shielded data (with Actions) test vectors
```
Comment (teor2345) [inline zebra-consensus/src/primitives/halo2/tests/vectors/orchard_shielded_data.rs]: Efficiency: Can you please put large test vectors in `zebra-test`, so they are only downloaded by test builds?

**Review by teor2345 — COMMENTED** [review]
I'd like us to keep large test vectors in `zebra-test`, so that they're not downloaded for release builds.

**Review by conradoplg — DISMISSED** [review]
Looks good! :+1: 

I'll leave unmerged so that you can address @teor2345's comments if needed

**Review by jvff — DISMISSED** [review]
Looks good! I left a few minor suggestions, but they're optional :)

File: zebra-consensus/Cargo.toml [hunk 17]
```diff
@@ -15,9 +15,12 @@ bellman = "0.10.0"
 bls12_381 = "0.5.0"
 chrono = "0.4.19"
 displaydoc = "0.2.2"
+halo2 = "=0.1.0-beta.1"
 jubjub = "0.7.0"
 lazy_static = "1.4.0"
 once_cell = "1.8"
+# TODO: replace with upstream orchard when these changes are merged
```
Comment (teor2345) [inline zebra-consensus/Cargo.toml:22]: Which changes?

Comment (dconnolly) [reply zebra-consensus/Cargo.toml:22]: Ah I noted in the top level `Cargo.toml` not this one, resolved in 0689c695

File: zebra-consensus/src/primitives/halo2/tests/vectors/orchard_shielded_data.rs (hunk 16 above)
Comment (dconnolly) [reply zebra-consensus/src/primitives/halo2/tests/vectors/orchard_shielded_data.rs]: Done!

2914acfa

File: deny.toml (hunk 11 above)
Comment (dconnolly) [reply deny.toml]: Done in cf4a5cabc6f87506b026a0680c3ef8c0bdf58e1a, thanks!

File: zebra-consensus/src/primitives/halo2/tests.rs (hunk 14 above)
Comment (dconnolly) [reply zebra-consensus/src/primitives/halo2/tests.rs]: Done in 4681f761

File: zebra-consensus/src/primitives/halo2/tests.rs (hunk 13 above)
Comment (dconnolly) [reply zebra-consensus/src/primitives/halo2/tests.rs]: Done in 4681f761

File: zebra-consensus/src/primitives/halo2/tests.rs (hunk 12 above)
Comment (dconnolly) [reply zebra-consensus/src/primitives/halo2/tests.rs]: Done in 4681f761

**Review by teor2345 — APPROVED** [review]
Looks good!
